### ISSUE TITLE
cpp: newlines added when nl_end_of_file = ignore

### ISSUE THREAD

**Issue by JodiTheTigger** [uncrustify/uncrustify#3863]
[//]: # "  If the issue is connected to existing Uncrustify options please, if possible, add the"
[//]: # "  following information to ease up the process:"
[//]: # "  • a link to a debug file:"
[//]: # "      generated with: 'uncrustify -p debug.txt -c pathToUsedConfig.cfg -f toBeFormatedFile.cpp' "
[//]: # "      Example: [debug.txt](https://linkToTheFile)"
[//]: # "      Example hosters for debug files: pastebin.com, gist.github.com, ..."
[//]: # "      The used config file is included in the debug file and does not need to be included here."
[//]: # 
[//]: # "  • include a small but complete test file that will be uncrustifyed"
[//]: # "  • include the generated results"
[//]: # "  • include the expected results"
[//]: # 
[//]: # "  ✋ please add a line containing ``` above and below of each of those three code sections"
[//]: # 
[//]: # "  • include the current version of your Uncrustify executable"
[//]: # "      printout via 'uncrustify -v'"
[//]: # "      Example: current version: uncrustify 0.63"
[//]: # "      or if possible additionally with the git sha of the commit"
[//]: # "                    current version: uncrustify 0.63 dc7b412"
[//]: # 
[//]: # "  • if possible include a version that worked"
[//]: # "      Example: working version: uncrustify 0.63"
[//]: # "      or"
[//]: # "                    working version: uncrustify 0.63 2a5e88f"

building off master, windows 11: `Uncrustify-0.75.0-356-5253a5ec7`

Input (note - there is _no trailing newline_)
```cpp
void MyClass::myFn()
{
}
```

cfg (note: ignore is the default anyway)
```
nl_end_of_file = ignore
```

expected
```
void MyClass::myFn()
{
}
```

got (notice: newline is added)
```
void MyClass::myFn()
{
}

```

```
# Uncrustify-0.75.0-356-5253a5ec7
# option(s) with 'not default' value: 0
#
# -=====-
# number of loops               = 1
# -=====-
# language                      = CPP
# -=====-
# Line                Tag         Parent_type  Type of the parent         Columns Br/Lvl/pp     Nl  Text
#   1>               TYPE|           FUNC_DEF|     PARENT_NOT_SET[  1/  1/  5/  0][0/0/0][0-0] void
#   1>               TYPE|           FUNC_DEF|     PARENT_NOT_SET[  6/  6/ 13/  1][0/0/0][0-0]      MyClass
#   1>          DC_MEMBER|           FUNC_DEF|     PARENT_NOT_SET[ 13/ 13/ 15/  0][0/0/0][0-0]             ::
#   1>           FUNC_DEF|               NONE|     PARENT_NOT_SET[ 15/ 15/ 19/  0][0/0/0][0-0]               myFn
#   1>        FPAREN_OPEN|           FUNC_DEF|     PARENT_NOT_SET[ 19/ 19/ 20/  0][0/0/0][0-0]                   (
#   1>       FPAREN_CLOSE|           FUNC_DEF|     PARENT_NOT_SET[ 20/ 20/ 21/  0][0/0/0][0-0]                    )
#   1>            NEWLINE|               NONE|     PARENT_NOT_SET[ 21/ 21/  1/  0][0/0/0][1-0]
#   2>         BRACE_OPEN|           FUNC_DEF|     PARENT_NOT_SET[  1/  1/  2/  0][0/0/0][0-0] {
#   2>            NEWLINE|               NONE|     PARENT_NOT_SET[  2/  2/  1/  0][1/1/0][1-0]
#   3>        BRACE_CLOSE|           FUNC_DEF|     PARENT_NOT_SET[  1/  1/  2/  0][0/0/0][0-0] }
#   3>            NEWLINE|               NONE|     PARENT_NOT_SET[  2/  1/  0/  0][0/0/0][1-0]
# -=====-
```

**Comment by JodiTheTigger** [comment]
Turned on all the logging, and this seems to happens in `steps.txt_002.log`

```
newline_add_between(741): and end->Text() is '}', orig line is 3, orig col is 1
   [CallStack:-DEBUG NOT SET-]
one_liner_nl_ok(3570): check type is BRACE_OPEN, parent is FUNC_DEF, flag is [EMPTY_BODY,PUNCTUATOR], orig line is 2, orig col is 1
one_liner_nl_ok(3575): true (not 1-liner), a new line may be added
newline_add_between(738): start->Text() is '{', type is BRACE_OPEN, orig line is 2, orig col is 1
newline_add_between(741): and end->Text() is '}', orig line is 3, orig col is 1
   [CallStack:-DEBUG NOT SET-]
one_liner_nl_ok(3570): check type is BRACE_OPEN, parent is FUNC_DEF, flag is [EMPTY_BODY,PUNCTUATOR], orig line is 2, orig col is 1
one_liner_nl_ok(3575): true (not 1-liner), a new line may be added
newlines_cleanup_braces(3943): orig line is 3, orig col is 1, Text() is '}'
log_rule(newlines_cleanup_braces:4367): rule is 'nl_brace_brace'
log_rule(newlines_cleanup_braces:4379): rule is 'nl_brace_square'
log_rule(newlines_cleanup_braces:4391): rule is 'nl_brace_fparen'
log_rule(newlines_cleanup_braces:4412): rule is 'nl_type_brace_init_lst_close'
log_rule(newlines_cleanup_braces:4422): rule is 'eat_blanks_before_close_brace'
log_rule(newlines_cleanup_braces:4479): rule is 'nl_brace_struct_var'
log_rule(newlines_cleanup_braces:4502): rule is 'nl_after_brace_close'
newline_end_newline(666): orig line is 3, orig col is 1, Text() is <Newline>
   type is NONE, parent type is NONE => new type is NEWLINE
newline_end_newline(671): 3:1 add newline after '}'
CopyAndAdd(A): orig line is 3, orig col is 1, <Newline>,
   @ after orig line is 3, orig col is 1, Text() is '}', type is BRACE_CLOSE,
   stage is OTHER [CallStack:-DEBUG NOT SET-]
newline_var_def_blk(2017): start orig line is 1, orig col is 1, Text() is 'void'
newline_var_def_blk(2040): orig line is 1, orig col is 1, Text() 'void', type is TYPE, PRE is false
newline_var_def_blk(2044): next_pc orig line is 1, orig col is 6, type is TYPE, Text() is 'MyClass'
```

note the line: `newline_end_newline(671): 3:1 add newline after '}'`

still investigating why that's happening

**Comment by JodiTheTigger** [comment]
I thought I found it - caused by not also checking for `IsNotNullChunk()` inside `newlines_cleanup_braces`. But that just fails a ton of c, cpp, and d unit tests :-(

`newlines.cpp`
```cpp
         {
            log_rule_B("nl_after_brace_close");
            Chunk *next = pc->GetNext();

            if (  next->IsNot(CT_SEMICOLON)
               && next->IsNot(CT_COMMA)
               && next->IsNot(CT_SPAREN_CLOSE)    // Issue #664
               && next->IsNot(CT_SQUARE_CLOSE)
               && next->IsNot(CT_FPAREN_CLOSE)
               && next->IsNot(CT_PAREN_CLOSE)
               && next->IsNot(CT_WHILE_OF_DO)
               && next->IsNot(CT_VBRACE_CLOSE)        // Issue #666
               && (  next->IsNot(CT_BRACE_CLOSE)
                  || !next->TestFlags(PCF_ONE_LINER)) // #1258
               && !pc->TestFlags(PCF_IN_ARRAY_ASSIGN)
               && !pc->TestFlags(PCF_IN_TYPEDEF)
               && !next->IsCommentOrNewline()
               && next->IsNotNullChunk())        // Fixes #3863 but causes c, cpp, d test failures
            {
               // #1258
               // dont add newline between two consecutive braces closes, if the second is a part of one liner.
               newline_end_newline(pc);
            }
```

I'll keep looking...

**Comment by JodiTheTigger** [comment]
Ok, gave up for now to figuring out a fix. However doing a `git bisect` has the regression caused by commit: 52e96033dac1b45f0b24819908f9fe8b90a25fa6 (`Complete replacement of chunk_is_not_token() with Chunk::IsNot()`)

**Comment by JodiTheTigger** [comment]
tl;dr: my orig fix is correct - just need to update unit tests as well as fix.

-------

ok, finally figured it out - the old code had a logic bug, and the new code didn't. However the new code didn't replicate the functionality of the code it replaced due to this broken code (it even updated all tests to expect a forced newline at the end of file where this happens)

(old) logic bug: when `IsNotNullChunk()` is false, then `chunk_is_not_token() != (!chunk_is_token())`

```cpp
static inline bool chunk_is_not_token(const Chunk *pc, E_Token c_token)
{
   return(  pc != nullptr
         && pc->IsNotNullChunk()
         && pc->IsNot(c_token));
}
```

```cpp
static inline bool chunk_is_token(const Chunk *pc, E_Token c_token)
{
   return(  pc != nullptr
         && pc->IsNotNullChunk()
         && pc->type == c_token);
}
```

so `chunk_is_not_token(next, CT_SEMICOLON)` would return `false` when `IsNotNullChunk()` is false, but `next->IsNot(CT_SEMICOLON)` returns `true` instead!

this means the old code had an implicit `&& IsNotNullChunk()` inside it
```cpp
            if (  chunk_is_not_token(next, CT_SEMICOLON)   // RAM: FALSE
               && chunk_is_not_token(next, CT_COMMA)   // RAM: FALSE
               && chunk_is_not_token(next, CT_SPAREN_CLOSE)    // Issue #664   // RAM: FALSE
               && chunk_is_not_token(next, CT_SQUARE_CLOSE)   // RAM: FALSE
               && chunk_is_not_token(next, CT_FPAREN_CLOSE)   // RAM: FALSE
               && chunk_is_not_token(next, CT_PAREN_CLOSE)   // RAM: FALSE
               && chunk_is_not_token(next, CT_WHILE_OF_DO)   // RAM: FALSE
               && chunk_is_not_token(next, CT_VBRACE_CLOSE) // Issue #666   // RAM: FALSE
               && (  chunk_is_not_token(next, CT_BRACE_CLOSE)
                  || !next->flags.test(PCF_ONE_LINER))      // #1258
               && !pc->flags.test(PCF_IN_ARRAY_ASSIGN)
               && !pc->flags.test(PCF_IN_TYPEDEF)
               && !next->IsCommentOrNewline())
```

so when replaced with (which doesn't)
```cpp
            if (  next->IsNot(CT_SEMICOLON)
               && next->IsNot(CT_COMMA)
               && next->IsNot(CT_SPAREN_CLOSE)    // Issue #664
               && next->IsNot(CT_SQUARE_CLOSE)
               && next->IsNot(CT_FPAREN_CLOSE)
               && next->IsNot(CT_PAREN_CLOSE)
               && next->IsNot(CT_WHILE_OF_DO)
               && next->IsNot(CT_VBRACE_CLOSE)         // Issue #666
               && (  next->IsNot(CT_BRACE_CLOSE)
                  || !next->flags.test(PCF_ONE_LINER)) // #1258
               && !pc->flags.test(PCF_IN_ARRAY_ASSIGN)
               && !pc->flags.test(PCF_IN_TYPEDEF)
               && !next->IsCommentOrNewline())
```

it's behaviour changed causing the regression.

**Comment by JodiTheTigger** [comment]
Wow, 24 hour turnaround! Cheers guys!
